File: poincare_clean/python_conversion/get_apar_sn.py
import numpy as np
from scipy.interpolate import CubicSpline
from scipy.fftpack import fft, ifft
import utils

# Define the function to calculate apar and its derivatives
def get_apar_sn(psi, bxy, psixy, zs, sa, sinty, dy0, dz,
                ixsep, nypf1, nypf2, zperiod, interp_opt, deriv_opt, true_apar):
    nx, ny, nz = psi.shape

    apar = np.zeros((nx, ny, nz))
    dapardx = np.zeros((nx, ny, nz))
    dapardy = np.zeros((nx, ny, nz))
    dapardz = np.zeros((nx, ny, nz))

    #ny_cfr = nypf2 - nypf1 + 1
    ny_cfr = nypf2 - nypf1
    print('ny_cfr= ', ny_cfr)
    print('nypf1,2= ', nypf1, nypf2)
    #iy_cfr = np.arange(nypf1, nypf2 + 1)
    iy_cfr = np.arange(nypf1, nypf2)
    #iy_pfr = np.concatenate([np.arange(nypf1), np.arange(nypf2 + 1, ny)])
    iy_pfr = np.concatenate([np.arange(nypf1), np.arange(nypf2, ny)])
    utils.write_array_to_file(iy_cfr, 'iy_cfr')
    utils.write_array_to_file(iy_pfr, 'iy_pfr')

    zarray = np.arange(nz) * dz
    zarrayp = np.arange(nz + 1) * dz
    zmax = dz * nz


    ## current code doesn't do this!!!
    if not true_apar:
        # Step 1: Interpolation of psi back to CELL_CENTER
        psis = psi.copy()
        psisp1 = np.zeros((nx, 1, nz))

        # Shift first point across the branch cut
        for i in range(ixsep):
            zarray_shift = (zarray + sa[i]) % zmax
            psis_tmp = psi[i, nypf1:, :].copy()
            psis_tmp = np.append(psis_tmp, psis_tmp[:, 0:1], axis=1)
            psisp1[i, 0, :] = CubicSpline(zarrayp, psis_tmp[:, 0])(zarray_shift)

        if interp_opt == 0:
            # Linear interpolation
            print("Linear interpolation of psi back to cell center.")
            for i in range(nx):
                if i <= ixsep:
                    # Closed flux surface region
                    psi[i, iy_cfr[:-1], :] = 0.5 * (psis[i, iy_cfr[:-1], :] + psis[i, iy_cfr[1:], :])
                    psi[i, nypf2, :] = 0.5 * (psisp1[i, 0, :] + psis[i, nypf2, :])
                    # Private flux region
                    psi[i, iy_pfr[:-1], :] = 0.5 * (psis[i, iy_pfr[:-1], :] + psis[i, iy_pfr[1:], :])
                    psi[i, iy_pfr[-1], :] = 1.5 * psis[i, iy_pfr[-1], :] - 0.5 * psis[i, iy_pfr[-2], :]
                else:
                    psi[i, :-1, :] = 0.5 * (psis[i, :-1, :] + psis[i, 1:, :])
                    psi[i, -1, :] = 1.5 * psis[i, -1, :] - 0.5 * psis[i, -2, :]

        elif interp_opt == 1:
            # Cubic spline interpolation
            print("Cubic spline interpolation of psi back to cell center.")
            for i in range(nx):
                if i <= ixsep:
                    for k in range(nz):
                        psis_cfr = psis[i, iy_cfr, k]
                        psis_cfr = np.append(psis_cfr, psisp1[i, 0, k])
                        psi[i, nypf1:nypf2, k] = CubicSpline(np.arange(1, ny_cfr + 1), psis_cfr)(np.arange(1, ny_cfr))

                        psis_pfr = psis[i, iy_pfr, k]
                        psis_pfr = np.append(psis_pfr, 2 * psis_pfr[-1] - psis_pfr[-2])
                        psi[i, iy_pfr, k] = CubicSpline(np.arange(1, 2 * nypf1 + 2), psis_pfr)(np.arange(1, 2 * nypf1 + 1))
                else:
                    for k in range(nz):
                        psis_tmp = psis[i, :, k]
                        psis_tmp = np.append(psis_tmp, psis_tmp[-1])
                        psi[i, :, k] = CubicSpline(np.arange(1, ny + 2), psis_tmp)(np.arange(1, ny + 1))

        else:
            raise ValueError("Unknown interpolation method for psi!")

        # Step 2: Calculate apar
        for k in range(nz):
            apar[:, :, k] = psi[:, :, k] * bxy

    else:
        # True apar
        apar = psi.copy()

    # Step 3: Calculate apar derivatives
    apars = np.zeros((nx, ny, nz))
    dapardpsi = np.zeros((nx, ny, nz))
    #kz=[0:nz/2 -nz/2+1:1:-1]*zperiod;
    # kz=kz';   ' is complex conjugate
    kz = np.concatenate([np.arange(0, nz//2 + 1), np.arange(-nz//2 + 1, 0)]) * zperiod
    kz = kz.reshape(-1, 1)  # Equivalent to transposing to a column vector
    utils.write_array_to_file(kz, 'kz')
    utils.write_array_to_file(zs, 'zs')

    ci = 1j

    for i in range(nx):
        for j in range(ny):
            ## apars(i,j,:)=real(ifft(fft(squeeze(apar(i,j,:))).*exp(-ci*zs(i,j)*kz)));
            # Apply FFT, multiply, and inverse FFT
            apar_ij = apar[i, j, :]  # Extract the (i, j, :) slice
            transformed = np.fft.fft(apar_ij) * np.exp(-ci * zs[i, j] * kz.flatten())
            apars[i, j, :] = np.real(np.fft.ifft(transformed))


    utils.write_array_to_file(apars, 'apars')
    shit()


    if deriv_opt == 0:
        # Numerical differentiation (finite difference)
        print("Central finite difference for Apar.")
        dpsi = np.abs(psixy[100, 38] - psixy[99, 38])
        dapardpsi[1:-1, :, :] = 0.5 * (apars[2:, :, :] - apars[:-2, :, :]) / dpsi

        dapardz[:, :, 0] = 0.5 * (apar[:, :, 1] - apar[:, :, -1]) / dz
        dapardz[:, :, 1:-1] = 0.5 * (apar[:, :, 2:] - apar[:, :, :-2]) / dz
        dapardz[:, :, -1] = 0.5 * (apar[:, :, 0] - apar[:, :, -2]) / dz

        for k in range(nz):
            dapardx[:, :, k] = dapardpsi[:, :, k] + sinty * dapardz[:, :, k]

        for i in range(nx):
            if i < ixsep:
                dapardy[i, iy_cfr[1:-1], :] = 0.5 * (apar[i, iy_cfr[2:], :] - apar[i, iy_cfr[:-2], :]) / dy0

                zarray_shift = (zarray - sa[i]) % zmax
                apar_tmp = apar[i, nypf2, :].copy()
                apar_tmp = np.append(apar_tmp, apar_tmp[0])
                aparm1 = CubicSpline(zarrayp, apar_tmp)(zarray_shift)
                dapardy[i, nypf1 + 1, :] = 0.5 * (apar[i, nypf1 + 2, :] - aparm1) / dy0

                zarray_shift = (zarray + sa[i]) % zmax
                apar_tmp = apar[i, nypf1, :].copy()
                apar_tmp = np.append(apar_tmp, apar_tmp[0])
                aparp1 = CubicSpline(zarrayp, apar_tmp)(zarray_shift)
                dapardy[i, nypf2, :] = 0.5 * (aparp1 - apar[i, nypf2 - 1, :]) / dy0

    elif deriv_opt == 1:
        # Analytical differentiation
        print("Cubic spline fit then differentiation of Apar.")
        for k in range(nz):
            for j in range(ny):
                apar_tmp = apars[:, j, k]
                apar_tmp = np.append(apar_tmp, apar_tmp[-1])
                xarray = psixy[:, j]
                xarray = np.append(xarray, 2 * xarray[-1] - xarray[-2])
                pp = CubicSpline(xarray, apar_tmp)
                dapardpsi[:, j, k] = pp(xarray, 1)

    return apar, dapardx, dapardy, dapardz

File: poincare_clean/python_conversion/get_apar_sn.m
% prepare and calcuate apar and its derivatives for field-line tracing
% for single-null configruation
%
% note that psi is defined at CELL_YLOW, while B field is at CELL_CENTER,
% so interpolation/shifting along y is needed UNLESS true_apar is true
function [apar,dapardx,dapardy,dapardz] = ...
    get_apar_sn(psi,bxy,psixy,zs,sa,sinty,dy0,dz, ...
    ixsep,nypf1,nypf2,zperiod,interp_opt,deriv_opt,true_apar)

    [nx,ny,nz]=size(psi);
    
    apar=zeros(nx,ny,nz);    dapardx=zeros(nx,ny,nz);
    dapardy=zeros(nx,ny,nz); dapardz=zeros(nx,ny,nz);
    ny_cfr=nypf2-nypf1+1;
    fprintf('ny_cfr= %d\n', ny_cfr);
    fprintf('nypf1,2= %d %d\n', nypf1, nypf2);


    iy_cfr=[nypf1+1:nypf2]; iy_pfr=[1:nypf1 nypf2+1:ny];
    zarray = (0:nz-1)*dz; zarrayp=(0:nz)*dz; zmax = dz*nz;

    write_array_to_file(iy_cfr, 'iy_cfr')
    write_array_to_file(iy_pfr, 'iy_pfr')

    if (~true_apar) % if not true_apar but psi as in most BOUT++ output

    % step 1, interpolation of psi back to to CELL_CENTER
    psis=psi;
    % shift first point across the branch-cut -- hence becomes ny+1 cell
    for i=1:ixsep
        zarray_shift=mod(zarray+sa(i),zmax);
        psis_tmp=squeeze(psi(i,nypf1+1,:)); psis_tmp(end+1)=psis_tmp(1);
        psisp1(i,1,:)=spline(zarrayp,psis_tmp,zarray_shift);
    end
    
    if (interp_opt == 0)
    % linear interpolation
    fprintf('Linear interpolation of psi back to cell center.\n');
    for i=1:nx
        if (i<ixsep+1)
            % closed flux surface region
            psi(i,iy_cfr(1:end-1),:)=0.5*(psis(i,iy_cfr(1:end-1),:)+psis(i,iy_cfr(2:end),:));
            psi(i,nypf2,:)=0.5*(psisp1(i,1,:)+psis(i,nypf2,:));
            % privite flux region
            psi(i,iy_pfr(1:end-1),:)=0.5*(psis(i,iy_pfr(1:end-1),:)+psis(i,iy_pfr(2:end),:));
            psi(i,iy_pfr(end),:)=1.5*psis(i,iy_pfr(end),:)-0.5*psis(i,iy_pfr(end-1),:); % simple linear extrapolation
        else
            psi(i,1:ny-1,:)=0.5*(psis(i,1:ny-1,:)+psis(i,2:ny,:));
            psi(i,ny,:)=1.5*psis(i,ny,:)-0.5*psis(i,ny-1,:); % simple linear extrapolation
        end
    end
    
    elseif (interp_opt == 1)
    % cubic spline interpolation
    fprintf('Cubic spline interpolation of psi back to cell center.\n');
    for i=1:nx
        if (i<ixsep+1)
            for k=1:nz
            % closed flux surface region
                psis_cfr=squeeze(psis(i,iy_cfr,k),k);
                psis_cfr(end+1)=psisp1(i,1,k);
                psi(i,nypf1+1:nypf2,k)=spline([1:ny_cfr],psis_cfr,[1:ny_cfr-1]+0.5);
            % privite flux region
                psis_pfr=squeeze(psis(i,iy_pfr,k));
                psis_pfr(end+1)=2.*psis_pfc(end)-psis_pfc(end-1);
                psi(i,iy_pfr,k)=spline([1:2*nypf1+1],psis_pfr,[1:2*nypf1]+0.5);
            end
        else
            for k=1:nz
            psis_tmp=psis(i,:,k); psis_tmp(end+1)=psis(end);
            psi(i,1:ny,k)=spline([1:ny+1],psis_tmp,[1:ny]+0.5);
            end
        end
    end
    else
        fprintf('Unknown interpolation method for psi!\n');
    end
    
    % step2, apar=psi*bxy unless psi is truly apar
    for k=1:nz
        apar(:,:,k)=psi(:,:,k).*bxy;
    end

    else % if the input is a true apar in the cell center
	    apar=psi;
    end

    % step 3, get apar derivatives
    
    % first shift to flux coordinate for d/dpsi
    apars=zeros(nx,ny,nz); dapardpsi=zeros(nx,ny,nz);
    kz=[0:nz/2 -nz/2+1:1:-1]*zperiod; kz=kz';

    write_array_to_file(kz, 'kz');
    write_array_to_file(zs, 'zs');
    ci=sqrt(-1);
    for i=1:nx
        for j=1:ny
            apars(i,j,:)=real(ifft(fft(squeeze(apar(i,j,:))).*exp(-ci*zs(i,j)*kz)));
        end
    end

    write_array_to_file(apars, 'apars');
    shit();
    
    if (deriv_opt == 0)
    % numerical differentiate, less accurate
    fprintf('Central finite difference for Apar.\n');
    dpsi=abs(psixy(101,39)-psixy(100,39));
    dapardpsi(2:nx-1,:,:)=0.5*(apars(3:nx,:,:)-apars(1:nx-2,:,:))/dpsi;

    dapardz(:,:,1)=0.5*(apar(:,:,2)-apar(:,:,nz))/dz;
    dapardz(:,:,2:nz-1)=0.5*(apar(:,:,3:nz)-apar(:,:,1:nz-2))/dz;
    dapardz(:,:,nz)=0.5*(apar(:,:,1)-apar(:,:,nz-1))/dz;
    
    for k=1:nz
        dapardx(:,:,k)=dapardpsi(:,:,k)+sinty.*dapardz(:,:,k);
    end
    
    % d/dy is a little tricky
    for i=1:nx
        if (i<ixsep+1)
            dapardy(i,iy_cfr(2:end-1),:)=0.5*(apar(i,iy_cfr(3:end),:)-apar(i,iy_cfr(1:end-2),:))/dy0;
            % reverse shift last point to be added before first point -- becomes -1 cell
            zarray_shift=mod(zarray-sa(i),zmax);
            apar_tmp=squeeze(apar(i,nypf2,:)); apar_tmp(end+1)=apar_tmp(1);
            aparm1(1,1,:)=spline(zarrayp,apar_tmp,zarray_shift);
            dapardy(i,5,:)=0.5*(apar(i,nypf1+2,:)-aparm1)/dy0;
            % then forward shift first point to be added to last point -- becomes ny+1 cell
            zarray_shift=mod(zarray+sa(i),zmax);
            apar_tmp=squeeze(apar(i,nypf1+1,:)); apar_tmp(end+1)=apar_tmp(1);
            aparp1(1,1,:)=spline(zarrayp,apar_tmp,zarray_shift);           
            dapardy(i,nypf2,:)=0.5*(aparp1-apar(i,nypf2-1,:))/dy0;
            % assume parallel gradient vanishes at divertor target
            dapardy(i,iy_pfr(2:end-1),:)=0.5*(apar(i,iy_pfr(3:end),:)-apar(i,iy_pfr(1:end-2),:))/dy0;
        else
            % again, parallel gradient vanishes at divertor target
            dapardy(i,2:ny-1,:)=0.5*(apar(i,3:ny,:)-apar(i,1:ny-2,:))/dy0;
        end
    end
    
    %or 4th order cfd (to be implemented)
    %dapardx(3:nx-2,:,:)=(-apar(5:nx,:,:)+8.*apar(4:nx-1,:,:)-8.*apar(2:nx-3)+apar(1:nx-4,:,:))/12./dx;
    
    elseif (deriv_opt == 1)
    % analytical differentiate based on cubic spline interpolation that
    % ensures first and second order derivatives are continous across the
    % cell boundaries
    fprintf('Cubic spline fit then differentiation of Apar.\n');
    
    for k=1:nz
        for j=1:ny
            apar_tmp=apars(:,j,k);
            apar_tmp(end+1)=apar_tmp(end); % pathing last point
            xarray=psixy(:,j);xarray(end+1)=2.*xarray(end)-xarray(end-1);
            pp=spline(xarray,apar_tmp);
            dapardpsi(:,j,k)=pp.coefs(:,3);
        end
    end    
    
    % d/dy again is a little tricky
    yiarray_sol=double(1:ny+1);yiarray_cfr=double(1:ny_cfr);
    for k=1:nz
        for i=1:nx
            if (i<ixsep+1)
                % closed flux surface region
                % forward shift first point to be added to last point -- becomes ny+1 cell
                zarray_shift=mod(zarray(k)+sa(i),zmax);
                apar_tmp=squeeze(apar(i,nypf1+1,:)); apar_tmp(end+1)=apar_tmp(1);
                aparp1=spline(zarrayp,apar_tmp,zarray_shift);
                apar_tmp=squeeze(apar(i,iy_cfr,k)); apar_tmp(end+1)=aparp1; % reuse apar_tmp, caution
                pp=spline(yiarray_cfr,apar_tmp);
                dapardy(i,iy_cfr,k)=pp.coefs(:,3);
                % private flux region
                apar_tmp=squeeze(apar(i,iy_pfr,k)); apar_tmp(end+1)=apar_tmp(end);
                pp=spline(double([1:2*nypf1+1]),apar_tmp);
                dapardy(i,iy_pfr,k)=pp.coefs(:,3);
            else % sol, easy~
                apar_tmp=squeeze(apar(i,:,k));
                apar_tmp(end+1)=apar_tmp(end);
                pp=spline(yiarray_sol,apar_tmp);
                dapardy(i,:,k)=pp.coefs(:,3);
            end
        end
    end
    dapardy=dapardy/dy0;
    
    for j=1:ny
        for i=1:nx
            apar_tmp=squeeze(apar(i,j,:));
            apar_tmp(end+1)=apar_tmp(1);
            pp=spline(zarrayp,apar_tmp);
            dapardz(i,j,:)=pp.coefs(:,3);
        end
    end
    
    for k=1:nz
        dapardx(:,:,k)=-dapardpsi(:,:,k)+sinty.*dapardz(:,:,k);
    end
    
    end
    
end
